### PR TITLE
Add type hints to data transformers

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Add type hints to data transformers
- Add type hints to util classes

### Fixed
- Allow null transformation in ModelsToArrayTransformer
```

### CODE REVIEW

File: src/Form/DataTransformer/ModelsToArrayTransformer.php [hunk 1]
```diff
@@ -125,6 +130,11 @@ public function __unset($name)
...
     }
 
+    /**
+     * @return string[]
+     *
+     * @phpstan-param T[] $collection
+     */
     public function transform($collection)
     {
         if (null === $collection) {
```
Comment (core23) [inline src/Form/DataTransformer/ModelsToArrayTransformer.php:140]: I'm not sure if we should check for invalid types in the data transformer classes.

Should we *trust* the typing system or should we check for every technical possible values.

Comment (core23) [reply src/Form/DataTransformer/ModelsToArrayTransformer.php:140]: Same for the other transfomers

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 2]
```diff
@@ -72,9 +80,14 @@ public function __construct(
         $this->toStringCallback = $toStringCallback;
     }
 
+    /**
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:83]: We could add a `@param`.

There is a check
```
if (!\is_array($value)) {
     throw new \UnexpectedValueException(sprintf('Value should be array, %s given.', \gettype($value)));
}
```

So `@param array` seems ok.

There is a check for empty values before the array check, 
but the symfony doc say
```
     * This method must be able to deal with empty values. Usually this will
     * be an empty string, but depending on your implementation other empty
     * values are possible as well (such as NULL). The reasoning behind
     * this is that value transformers must be chainable. If the
     * reverseTransform() method of the first value transformer outputs an
     * empty string, the second value transformer must be able to process that
     * value.
```

So it would means that allowed values are arrays, `null` and `''` ?

File: src/Form/DataTransformer/ModelsToArrayTransformer.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelsToArrayTransformer.php:140]: As explained here: https://github.com/sonata-project/SonataAdminBundle/pull/6361#discussion_r483943766

The Phpdoc from Symfony is saying that we must deal empty value, which is generally `null`.

What I understand is that empty values are valid values.
So it's not about trust the typing system, it's about fixing in.

Phpdoc should be `@phpstan-param T[]|null $collection` because the `null` case can occur.

File: src/Form/DataTransformer/ArrayToModelTransformer.php [hunk 3]
```diff
@@ -59,6 +68,9 @@ public function reverseTransform($array)
         return $this->modelManager->modelReverseTransform($this->className, $array);
     }
 
+    /**
+     * @phpstan-param T $value
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ArrayToModelTransformer.php:72]: For the same reason than https://github.com/sonata-project/SonataAdminBundle/pull/6377/files#r483949650,
I feel like the phpdoc should be `T|null`.

And we should also add `@param object|null $value`

File: src/Form/DataTransformer/ModelsToArrayTransformer.php [hunk 4]
```diff
@@ -141,13 +151,18 @@ public function transform($collection)
...
     }
 
+    /**
+     * @return Collection<int|string, object>
+     *
+     * @phpstan-return Collection<array-key, T>
+     */
     public function reverseTransform($keys)
     {
         if (!\is_array($keys)) {
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelsToArrayTransformer.php:161]: Similar to https://github.com/sonata-project/SonataAdminBundle/pull/6377/files#r483949339,

The Symfony phpdoc say that we must be able to deal with empty values, should we add a check then ?

File: src/Form/DataTransformer/ModelsToArrayTransformer.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelsToArrayTransformer.php:140]: I like the way it's done here: 
https://github.com/symfony/form/blob/master/Extension/Core/DataTransformer/StringToFloatTransformer.php
Or 
https://github.com/symfony/form/blob/master/Extension/Core/DataTransformer/ChoicesToValuesTransformer.php

Comment (core23) [reply src/Form/DataTransformer/ModelsToArrayTransformer.php]: I changed this to always except null values for both transformations.

File: src/Form/DataTransformer/ArrayToModelTransformer.php [hunk 5]
```diff
@@ -30,35 +32,57 @@ class ArrayToModelTransformer implements DataTransformerInterface
...
+        if (!\is_array($value)) {
             return $instance;
         }
 
-        return $this->modelManager->modelReverseTransform($this->className, $array);
+        return $this->modelManager->modelReverseTransform($this->className, $value);
     }
 
+    /**
+     * @param mixed|null $value
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ArrayToModelTransformer.php:78]: If you use `@phpstan-param T|null $value`, you should use `@param object|null $value`

Same for `@return`.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 6]
```diff
@@ -72,9 +80,16 @@ public function __construct(
...
+     * @param mixed[]|null $value
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: Why phpstan/psalm doesn't report that 
```
if (!\is_array($value)) {
    throw new \UnexpectedValueException(sprintf('Value should be array, %s given.', \gettype($value)));
}
```
is useless ?

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 7]
```diff
@@ -104,25 +119,34 @@ public function reverseTransform($value)
...
                     .' Make sure that form option "multiple=false" is set for many-to-one relation and "multiple=true"'
                     .' is set for many-to-many or one-to-many relations.'
                 );
             }
-            if ($isArray || ($entityOrCollection instanceof \ArrayAccess)) {
-                $collection = $entityOrCollection;
+            if ($isArray || ($value instanceof \ArrayAccess)) {
+                $collection = $value;
             } else {
                 throw new \InvalidArgumentException(
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:151]: There could be some improvements in the check to avoid throwing the same exception here and in the previous `if`.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 8]
```diff
@@ -131,16 +155,16 @@ public function transform($entityOrCollection)
                 );
             }
         } else {
-            if (substr(\get_class($entityOrCollection), -1 * \strlen($this->className)) === $this->className) {
-                $collection = [$entityOrCollection];
-            } elseif ($entityOrCollection instanceof \ArrayAccess) {
+            if (substr(\get_class($value), -1 * \strlen($this->className)) === $this->className) {
+                $collection = [$value];
+            } elseif ($value instanceof \ArrayAccess) {
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:160]: shouldn't it be `\is_array($value) || $value instanceof \ArrayAccess` ?

So it seems like we doing the same checks in the `this->multiple` case and the not multiple one. We could definitely improve the conditions.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 9]
```diff
@@ -131,16 +155,16 @@ public function transform($entityOrCollection)
                 );
             }
         } else {
-            if (substr(\get_class($entityOrCollection), -1 * \strlen($this->className)) === $this->className) {
-                $collection = [$entityOrCollection];
-            } elseif ($entityOrCollection instanceof \ArrayAccess) {
+            if (substr(\get_class($value), -1 * \strlen($this->className)) === $this->className) {
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:158]: It should be `!$isArray && substr(\get_class($value), -1 * \strlen($this->className)) === $this->className`

File: src/Form/DataTransformer/ModelToIdTransformer.php [hunk 10]
```diff
@@ -30,33 +32,51 @@ class ModelToIdTransformer implements DataTransformerInterface
...
      */
     public function __construct(ModelManagerInterface $modelManager, $className)
     {
         $this->modelManager = $modelManager;
         $this->className = $className;
     }
 
-    public function reverseTransform($newId)
+    /**
+     * @param mixed|null $value
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdTransformer.php:52]: Null is included in mixed.

```suggestion
     * @param mixed $value
```

File: src/Form/DataTransformer/ModelToIdTransformer.php [hunk 11]
```diff
@@ -30,33 +32,51 @@ class ModelToIdTransformer implements DataTransformerInterface
...
 
-        return $this->modelManager->find($this->className, $newId);
+        return $this->modelManager->find($this->className, $value);
     }
 
-    public function transform($model)
+    /**
+     * @param object|null $value
+     *
+     * @return mixed|null
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ModelToIdTransformer.php:70]: I think that `getNormalizedIdentifier` return `string`.

```suggestion
     * @return string|null
```

**Comment by core23** [comment]
Backported type hints from https://github.com/sonata-project/SonataAdminBundle/pull/6366

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 6 above)
Comment (jordisala1991) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: because it could be null

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 7 above)
Comment (core23) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:151]: This method is also very long and should be refactored. This should be done in another PR.

File: src/Form/DataTransformer/ArrayToModelTransformer.php [hunk 12]
```diff
@@ -30,35 +32,57 @@ class ArrayToModelTransformer implements DataTransformerInterface
...
      */
     public function __construct(ModelManagerInterface $modelManager, $className)
     {
         $this->modelManager = $modelManager;
         $this->className = $className;
     }
 
-    public function reverseTransform($array)
+    /**
+     * @param object|array[]|null $value
```
Comment (VincentLanglet) [inline src/Form/DataTransformer/ArrayToModelTransformer.php:52]: Are you sure it can be an array of array ?

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 6 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: There is a `if (empty($value)) {` made earlier

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:151]: Yes, I didn't considered this as a blocking point for the merge.

**Comment by jordisala1991** [comment]
I dont recommend backporting from an unfinished PR 😅.

File: src/Form/DataTransformer/ArrayToModelTransformer.php (hunk 12 above)
Comment (core23) [reply src/Form/DataTransformer/ArrayToModelTransformer.php]: Good catch!

**Comment by core23** [comment]
> I dont recommend backporting from an unfinished PR 😅.

I did not only backport the hints, I merged and verified some of your changes.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 6 above)
Comment (jordisala1991) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: Hum, but empty is more like to test if an array is empty.

Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: Sure, but what I wanted to said if that `$value` can't be null during the 
```
if (!\is_array($value)) {
```
Check, since the `null` value are already caught by the
```
if (empty($value)) {
```
Check.

Either we should use `@param mixed $value` either we could remove the array check.

Comment (core23) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: IMHO we should keep the array check, because phpstan attributes does not enforce valid types.

File: src/Form/DataTransformer/ModelsToArrayTransformer.php (hunk 4 above)
Comment (core23) [reply src/Form/DataTransformer/ModelsToArrayTransformer.php]: Fixed it

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 6 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:84]: > IMHO we should keep the array check, because phpstan attributes does not enforce valid types.

I'm not against this check. But then, it would be better to use `@param mixed $value`.

If one day we change 
```
if (empty($value)) {
```
to 
```
if (null === $value) {
```
We would end with a phpstan error with our current set up.

Comment (core23) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php]: Changed it back to `mixed`

**Review by VincentLanglet — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thank you @core23
